### PR TITLE
Breadcrumb: Fix automation attribute?

### PR DESCRIPTION
**Pull request by siddharthkp**
Sorry I wrote an essay

tl;dr: found a bug, fix means breaking change, we might be able to squeeze this in in without breaking anything

&nbsp;

---

&nbsp;

### Automation attribute bug

&nbsp;

When you render

```jsx
<Breadcrumb>
  <Breadcrumb.Link>Home</Breadcrumb.Link>
  <Breadcrumb.Link>Parent</Breadcrumb.Link>
</Breadcrumb>
```

These are the cosmos-key it renders:

```html
<div data-cosmos-key="breadcrumb">
  <a data-cosmos-key="link">Home</a>
  <a data-cosmos-key="link">Parent</a>
</div>
```

There are 2 problems with the `data-cosmos-key` of the child being `link`:

1. It seems to rely on an implementation detail - the breadcrumb uses a `Link` component today, this might change (maybe to `<Button appearance="link">` 🤷‍♀️)

2. This is not what the code says! The code says `breadcrumb.link`

```jsx
const Breadcrumb.Link = props =>  <Link {...props} {...Automation('breadcrumb.link')} />
//                                                                          ⬆️
```

Then how does this happen 🤔 

---

&nbsp;

### A tale why order of props is important

&nbsp;

When you render 

```jsx
const Link = (props) => <a href="a.com" href="b.com">clicky</a>

render(<Link/>)
```

it results in:

```html
<a href="b.com">clicky</a>
```

**The second prop wins.**

&nbsp;

Using objects makes this more difficult to see:

```jsx
// breadcrumb.link wants to attach it's own cosmos-key
const Breadcrumb.Link = (props) => <Link data-cosmos-key="breadcrumb.link">{props.children}</Link>

// we've got ...props and link's own cosmos-key
const Link = (props) =>  <a {...props} data-cosmos-key="link">{props.children}</a>

render(<Breadcrumb.Link>clicky</Breadcrumb.Link>)
```

this will render

```html
<a data-cosmos-key="link">clicky</a>
```

Whoops, you'd want the key to be `breadcrumb.link` but the **second prop won.**

```jsx
const Link = (props) =>  <a {...props} data-cosmos-key="link">{props.children}</a>
//                               1️⃣                 2️⃣
```

&nbsp;

The fix is pretty straightforward:

```diff
- const Link = (props) =>  <a {...props} data-cosmos-key="link">clicky</a>
+ const Link = (props) =>  <a data-cosmos-key="link" {...props}>clicky</a>
//                               1️⃣                      2️⃣
```

now it renders

```html
<a data-cosmos-key="breadcrumb.link">clicky</a>
// 👍 
```

&nbsp;

This is a useful feature of React but also easy to screw up. Double edged swords? ⚔️ 

----

&nbsp;

### Guidelines for prop

(added this to contribution guide)

&nbsp;

1. **Components should respect the props passed to them and set them on the underlying element**

    ```jsx
    const Link = (props) => <a {...props}>clicky</a>

    render(<Link id="custom-id">)
    ```

    renders 

    ```html
    <a id="custom-id">
    ```

    This is how the new `margin` prop works as well 😉 

&nbsp;

2. **Props should override local attributes.**

    This will make sure the parent gets preference over what is rendered.

    When you add an automation attribute in `Breadcrumb.Link`, you are overriding the automation attribute of the underlying `Link`

	 ```jsx
    const Breadcrumb.Link = (props) => (
  	    <Link {...Automation('breadcrumb.link')}>{props.children}</Link>
    )

    const Link = (props) => <a {...Automation('link')} {...props}>clicky</a>
	```

    This way, the user of the component (Breadcrumb here) gets the final say in what is rendered.

    ```html
    <a data-cosmos-key="breadcrumb.link"/>
    ```

	This principle extends to cosmos users as well, they can give their own attributes to cosmos components.

    ```jsx
    <Link rel="nofollow" />
    ```

    or even

    ```jsx
    <Link data-cosmos-key="custom-component.link" />
	```

	&nbsp;

    This is usually the behaviour you want from any atomic component. So it's useful to make this into a rule.

    **`{...props}` should be the last thing inside a Component.Element**

&nbsp;

3. **When needed, cosmos can override the user's intent**

	This is a tricky one, you usually want the user to be able to have the final say.

    But, in some cases, you can override them to save them from making a mistake (usually a design mistake)

	Example:

    ```jsx
    const InputAction = (props) => <Button {...props} appearance="link" />
    ```

    Here we are locking `appearance` to `link` even if the user passed `appearance="destructive"`

&nbsp;
   
----

&nbsp;

Cool story bro, let's merge this?

&nbsp;

### Breaking change 🙈

&nbsp;

Fun fact, [our docs](https://auth0-cosmos.now.sh/docs/#/automation) documents the bug 😄 

Funner fact, the automation tests did not fail when I made this breaking change 🙈 

```
Breadcrumb Link: $('div[data-cosmos-key="breadcrumb"] a[data-cosmos-key="link"]')
//                                                                        ⬆️🙃 
```

Even though the fix is pretty simple, it is breaking in nature. It will change snapshots and fail automation tests.

Change in snapshots is acceptable as it is part of the implementation detail that is expected to change.

Automation tests failing with a minor upgrade is however not expected.

**The question here is: can we get away by making this change and hope nobody notices?**

&nbsp;

----

&nbsp;

### This must be happening everywhere!

&nbsp;

Yeeep. Any component that doesn't follow those guidelines ⬆️ will have such subtle bugs.

But that's okay, we'll end up fixing all of them as part of https://github.com/auth0/cosmos/pull/1308 and https://github.com/auth0/cosmos/issues/1270 as they keep surfacing.

I'm glad we took up the tasking task in time. kudos to @francocorreasosa for championing this 👏 

&nbsp;

### CODE REVIEW

**Comment by francocorreasosa** [comment]
Very nice explanation @siddharthkp 👏

> The question here is: can we get away by making this change and hope nobody notices?

Nope, I don't think so. But, I think this is something we need to fix sooner than later, we should make a list of changes in automation attributes so it's easier for folks to update their tests. Snapshots are the easiest, but this attributes are meant to be used in E2E tests, where they are going to update those values by hand.

### How can we prevent this from happening again?

I am wondering if we could create an extra helper for this. Example 
```jsx
<Component {...ApplyProps({ props, automation: 'breadcrumb.link' })} />
``` 

So we always ensure they are passed in the right order. (Internally we merge those objects correctly (and more important, **consistently**).

Let me know what you think

**Comment by siddharthkp** [comment]
> ### How can we prevent this from happening again?
>
> I am wondering if we could create an extra helper for this

I'm leaning more towards a linter than a helper. Something that can enforce a convention

Example: These 3 should come in a certain order whenever found:

```jsx
const overrides = { appearance: 'link' }

const InputAction = props => (
  <Button {...Automation('button')} {...props} {...overrides} />
)
```

**Comment by francocorreasosa** [comment]
A linter sounds good as well. Let's prioritize it in the next planning 💪

**Comment by siddharthkp** [comment]
#1308 is merged, now we need to decide what do we do with this PR

**Comment by siddharthkp** [comment]
We have a 👍 to merge
